File: plugins/spigot/TitleManager/src/main/java/io/puharesource/mc/titlemanager/commands/TMSubCommand.java
package io.puharesource.mc.titlemanager.commands;

import org.bukkit.ChatColor;
import org.bukkit.command.CommandSender;

import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import io.puharesource.mc.titlemanager.backend.language.Messages;
import lombok.Getter;

public abstract class TMSubCommand {

    private @Getter final String alias;
    private @Getter final String node;
    private @Getter final String usage;
    private @Getter final String description;
    private @Getter final String[] aliases;

    private final Set<CommandParameterIdentifier> supportedParameters;

    public TMSubCommand(final String alias, final String node, final Messages usage,
                        final Messages description, final String... aliases) {
        this.alias = alias;
        this.node = node;
        this.usage = usage.getMessage();
        this.description = description.getMessage();
        this.aliases = aliases;
        this.supportedParameters = new HashSet<>();

        if (this.getClass().isAnnotationPresent(ParameterSupport.class)) {
            ParameterSupport parameterSupport = this.getClass().getAnnotation(ParameterSupport.class);
            Collections.addAll(supportedParameters, parameterSupport.supportedParams());
        }
    }

    public abstract void onCommand(final CommandSender sender, final String[] args,
                                   final CommandParameters params) throws TMCommandException;

    public void syntaxError(final CommandSender sender) {
        sender.sendMessage(ChatColor.RED + "Wrong usage! Correct usage:");
        sender.sendMessage(ChatColor.RED + "   /" + getAlias() + " " + getUsage());
    }

    public Collection<CommandParameterIdentifier> getSupportedParameters() {
        return supportedParameters;
    }

    private void sendFormattedMessage(final boolean silent, final CommandSender sender,
                                      final ChatColor color, final Messages message,
                                      final Object... args) {

        if (silent) return;

        sender.sendMessage(color + String.format(message.getMessage()
                .replace("%s", ChatColor.RESET + "%s" + color), args));
    }

    protected void sendSuccess(final boolean silent, final CommandSender sender,
                               final Messages message, final Object... args) {

        sendFormattedMessage(silent, sender, ChatColor.GREEN, message, args);
    }

    protected void sendError(final boolean silent, final CommandSender sender,
                             final Messages message, final Object... args) {

        sendFormattedMessage(silent, sender, ChatColor.RED, message, args);
    }
}

File: src/main/kotlin/io/puharesource/mc/titlemanager/internal/commands/TMSubCommand.kt
package io.puharesource.mc.titlemanager.internal.commands

import io.puharesource.mc.titlemanager.TitleManagerPlugin
import io.puharesource.mc.titlemanager.internal.extensions.color
import io.puharesource.mc.titlemanager.internal.extensions.stripColor
import io.puharesource.mc.titlemanager.internal.model.command.AllowedCommandSender
import io.puharesource.mc.titlemanager.internal.model.command.CommandExecutor
import io.puharesource.mc.titlemanager.internal.model.command.CommandParameter
import net.md_5.bungee.api.ChatColor
import org.bukkit.command.CommandSender
import java.util.concurrent.Executor

class TMSubCommand constructor(
    val plugin: TitleManagerPlugin,
    val name: String,
    val allowedSender: AllowedCommandSender = AllowedCommandSender.ALL,
    private val executor: Executor? = null,
    val aliases: Set<String> = setOf(),
    val cmdExecutor: (TMSubCommand, CommandSender, Array<out String>, Map<String, CommandParameter>) -> CommandExecutor
) {
    fun runCommand(sender: CommandSender, args: Array<out String>, parameters: Map<String, CommandParameter>) {
        if (executor == null) {
            cmdExecutor(this, sender, args, parameters)
        } else {
            executor.execute { cmdExecutor(this, sender, args, parameters) }
        }
    }

    fun syntaxError(sender: CommandSender) {
        sender.sendMessage("${ChatColor.RED}Wrong usage! Correct usage:")

        sender.sendMessage(
            buildString {
                append("${ChatColor.RED}").append("    /tm ").append(name.lowercase())

                if (usage.isNotBlank()) {
                    append(" ").append(usage)
                }

                // If a description is specified, add the description to the message.
                if (description.isNotBlank()) {
                    append("${ChatColor.GRAY}").append(" - ").append(description)
                }
            }
        )
    }

    val permission: String
        get() = "titlemanager.command.$name"

    val usage: String
        get() = plugin.config.getString("messages.command-$name.usage").orEmpty().color().stripColor()

    val description: String
        get() = plugin.config.getString("messages.command-$name.description").orEmpty().color().stripColor()
}
